source/extensions/filters/http/oauth2/filter.cc

```diff
@@ -470,19 +459,38 @@ void OAuth2Filter::finishFlow() {
       Hex::encode(crypto_util.getSha256Hmac(token_secret_vec, token_payload));
   std::string encoded_token;
   absl::Base64Escape(pre_encoded_token, &encoded_token);
+  return encoded_token;
+}
 
-  // We use HTTP Only cookies for the HMAC and Expiry.
-  const std::string cookie_tail = fmt::format(CookieTailFormatString, new_expires_);
-  const std::string cookie_tail_http_only =
-      fmt::format(CookieTailHttpOnlyFormatString, new_expires_);
+void OAuth2Filter::onGetAccessTokenSuccess(const std::string& access_code,
+                                           const std::string& id_token,
+                                           const std::string& refresh_token,
+                                           std::chrono::seconds expires_in) {
+  updateTokens(access_code, id_token, refresh_token, expires_in);
+  finishGetAccessTokenFlow();
+}
 
+void OAuth2Filter::finishGetAccessTokenFlow() {
   // At this point we have all of the pieces needed to authorize a user.
   // Now, we construct a redirect request to return the user to their
   // previous state and additionally set the OAuth cookies in browser.
   // The redirection should result in successfully passing this filter.
-  Http::ResponseHeaderMapPtr response_headers{Http::createHeaderMap<Http::ResponseHeaderMapImpl>(
-      {{Http::Headers::get().Status, std::to_string(enumToInt(Http::Code::Found))}})};
+  Http::ResponseHeaderMapPtr response_headers = fillResponseHeader(getEncodedToken());
+  response_headers->setStatus(std::to_string(enumToInt(Http::Code::Found)));
+  response_headers->setLocation(state_);
 
+  decoder_callbacks_->encodeHeaders(std::move(response_headers), true, REDIRECT_LOGGED_IN);
+  config_->stats().oauth_success_.inc();
+}
+
+Http::ResponseHeaderMapPtr
+OAuth2Filter::fillResponseHeader(const std::string& encoded_token) const {
```

Comment:
Maybe it would be a bit clearer if this function was called `addResponseCookies` and it took a `&ResponseHeaderMap` that it populates with the cookies? That way we can keep the previous 
```
 Http::ResponseHeaderMapPtr response_headers{Http::createHeaderMap<Http::ResponseHeaderMapImpl>(
      {{Http::Headers::get().Status, std::to_string(enumToInt(Http::Code::Found))}})};
```
init of the header map and use this helper function to add all the cookies?